src/sentry/models/organization.py

```diff
@@ -272,6 +290,46 @@ def update(self, *args, **kwargs):
     def reserve_snowflake_id(cls):
         return generate_snowflake_id(cls.snowflake_redis_key)
 
+    # Used to generate Organization specific slugs with a region prefix if one
+    # is required. If the slug is prefixed with any character combination
+    # resulting in it starting with `r-XX-`, this sequence will be truncated
+    # by default in order to prevent a collision with any future regions.
+    @classmethod
+    def generate_regional_slug(
+        cls, slugify_target: str, truncate_region_prefix_collisions: bool = True
+    ) -> str:
+        sanitized_slug_target = slugify_target.lower().replace("_", "-").strip("-")
+        starts_with_region_prefix = match(region_prefix_regex, sanitized_slug_target)
+
+        if starts_with_region_prefix and truncate_region_prefix_collisions:
+            region_mod_split = split(region_prefix_regex, sanitized_slug_target)
+            sanitized_slug_target = region_mod_split[2]
+
+        elif starts_with_region_prefix:
+            raise InvalidRegionalSlugTargetException(
+                "Cannot slugify a slug that contains a region prefix"
+            )
+
+        if not sanitized_slug_target:
+            raise InvalidRegionalSlugTargetException(
+                "Slugify target specified either empty or only contains region prefixes"
+            )
+
+        local_region = get_local_region()
+        if local_region.name != settings.SENTRY_MONOLITH_REGION:
+            sanitized_slug_target = f"r-{local_region.name}-{sanitized_slug_target}"
+
+        # Because slugification is currently tied to model instances, we're just creating a dummy
+        # org to generate a slug
+        slug_generator_org = cls()
+
+        lock = locks.get("slug:organization", duration=5, name="organization_slug")
+        with TimedRetryPolicy(10)(lock.acquire):
+            slugify_instance(
+                slug_generator_org, sanitized_slug_target, reserved=RESERVED_ORGANIZATION_SLUGS
+            )
+        return slug_generator_org.slug
```

Comment:
Originally, I wrote this method as part of the Organization actions domain functions file, but hit some circular imports because of it.

Instead of having to inline all of the organization imports there, I decided to just pull it into the Org class itself, but this feels like a mixing of concerns tbh 🤔